### PR TITLE
Improvements/#269 nccl optimization

### PR DESCRIPTION
**Pull request by mhuguesaws**
*Issue #, if available:*
#269 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile [hunk 1]
```diff
@@ -3,46 +3,54 @@
 FROM nvidia/cuda:12.2.0-devel-ubuntu22.04
 
 ARG EFA_INSTALLER_VERSION=1.31.0
-ARG AWS_OFI_NCCL_VERSION=v1.8.1-aws
+ARG AWS_OFI_NCCL_VERSION=1.8.1
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:6]: why drop the `aws`? I was keeping this the same as the Github release tag to make it easy to see which versions are available, just by looking at https://github.com/aws/aws-ofi-nccl/releases

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile [hunk 2]
```diff
@@ -3,46 +3,54 @@
...
 ARG NCCL_TESTS_VERSION=2.13.9
 ARG NCCL_VERSION=2.20.3
+ARG GDRCOPY_VERSION=2.4.1
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:9]: do we need GDRCopy for all instance types?

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile (hunk 1 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:6]: To make simple simple and consistent with other args version value.

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile (hunk 2 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:9]: yes

File: micro-benchmarks/nccl-tests/README.md [hunk 3]
```diff
@@ -35,119 +35,171 @@ The NCCL tests are packaged in a container.
...
 ```
 
+To change the type of collective to test, modify the line with `srun` in the file `nccl-tests.sbatch` and change `all_reduce_perf` to any of: `all_gather_perf`, `alltoall_perf`, `gather_perf`, `reduce_perf`, `scatter_perf`, `all_reduce_perf`, `broadcast_perf`, `hypercube_perf`, `reduce_scatter_perf`, `sendrecv_perf`.
 
-### 2.2 Validate the NCCL configuration
-
-You can validate your environment for NCCL using the batch file `3.nccl-validate.sbatch`. Submit it as follows:
-
-```bash
-sbatch 3.nccl-validate.sbatch
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/README.md:144]: Why did we remove this?

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile (hunk 1 above)
Comment (sean-smith) [reply micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:7]: I'd prefer to use the full release tag

File: micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch [hunk 4]
```diff
@@ -1,10 +1,14 @@
 #!/bin/bash
-#SBATCH -N 2
-#SBATCH --exclusive
-
 # Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 # SPDX-License-Identifier: MIT-0
 
+#SBATCH --job-name=nccl-all_reduce_perf # name of your job
+#SBATCH --nodes=2 # number of nodes to use, 24 p4d(e) = 192 A100 GPUs
+#SBATCH --ntasks-per-node 8 # Number of GPU per node
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch:7]: We need to check if this works on SMHP

File: micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch [hunk 5]
```diff
@@ -16,10 +20,12 @@ mpirun -N 1 bash -c 'echo $(hostname) ➡️ $(cat /sys/devices/virtual/dmi/id/b
 mpirun -n $((8 * SLURM_JOB_NUM_NODES)) -N 8 \
         -x FI_PROVIDER=efa \
 	-x FI_EFA_USE_DEVICE_RDMA=1  \
-	-x RDMAV_FORK_SAFE=1 \
+	-x FI_EFA_FORK_SAFE=1 \
 	-x LD_LIBRARY_PATH=/opt/amazon/efa/lib:/opt/amazon/openmpi/lib:/opt/aws-ofi-nccl/lib:/usr/local/lib:/usr/lib:$LD_LIBRARY_PATH \
 	-x NCCL_DEBUG=INFO \
-	--mca pml ^cm \
+	-x NCCL_BUFFSIZE=8388608 \
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch:26]: Can we explain why we include these flags?

File: micro-benchmarks/nccl-tests/README.md [hunk 6]
```diff
@@ -35,119 +35,171 @@ The NCCL tests are packaged in a container.
...
 
-### Slurm
-
-To run the NCCL tests on Slurm, you will need to build the container then convert it into a Squash file using Enroot.
-
-To build the container:
-
-1. Copy the file `0.nccl-tests.Dockerfile` or its content to your head-node.
-2. Build the container image with the command below
+### Build the container
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/README.md:43]: Can you include a section title `## Deep Learning AMI`

File: micro-benchmarks/nccl-tests/README.md [hunk 7]
```diff
@@ -35,119 +35,171 @@ The NCCL tests are packaged in a container.
 
 > | Variable              | Default     |
 > |-----------------------|-------------|
-> |`EFA_INSTALLER_VERSION`| `latest`    |
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/README.md:38]: Link to respective libraries

File: micro-benchmarks/nccl-tests/slurm/nccl-tests.sbatch [hunk 8]
```diff
@@ -3,10 +3,12 @@
 # Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 # SPDX-License-Identifier: MIT-0
 
-#SBATCH -N 2 # number of nodes to use, 24 p4d(e) = 192 A100 GPUs
-#SBATCH --job-name=megatron_gpt # name of your job
+#SBATCH --job-name=nccl-all_reduce_perf # name of your job
+#SBATCH --nodes=2 # number of nodes to use, 24 p4d(e) = 192 A100 GPUs
 #SBATCH --ntasks-per-node 8 # Number of GPU per node
-#SBATCH --gres=gpu:8 # number of GPU we reserve
+#SBATCH --gpus-per-node=8 # number of GPU we reserve
```
Comment (sean-smith) [inline micro-benchmarks/nccl-tests/slurm/nccl-tests.sbatch:9]: This will break SMHP

File: micro-benchmarks/nccl-tests/nccl-tests.Dockerfile (hunk 1 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/nccl-tests.Dockerfile:7]: will update

File: micro-benchmarks/nccl-tests/README.md (hunk 3 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/README.md:144]: Not needed. It's redundant.

File: micro-benchmarks/nccl-tests/slurm/nccl-tests.sbatch (hunk 8 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/slurm/nccl-tests.sbatch:9]: We can comment the `gres` line

File: micro-benchmarks/nccl-tests/README.md (hunk 7 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/README.md:38]: Link meaning repository links?

File: micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch (hunk 5 above)
Comment (mhuguesaws) [reply micro-benchmarks/nccl-tests/slurm/nccl-tests-deep-learning-ami.sbatch:26]: They are explained in the `nccl-tests.sbatch` I can duplicate comment in there.

File: micro-benchmarks/nccl-tests/README.md (hunk 7 above)
Comment (sean-smith) [reply micro-benchmarks/nccl-tests/README.md:38]: yes, as in links to the respective release pages, i.e.

```
> | [EFA_INSTALLER_VERSION](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html#efa-start-enable) | `1.31.0`    |
> | [AWS_OFI_NCCL_VERSION](https://github.com/aws/aws-ofi-nccl/releases) | `1.8.1`     |
> | [NCCL_VERSION](https://docs.nvidia.com/deeplearning/nccl/release-notes/index.html)         | `2.20.3`    |
> | [NCCL_TESTS_VERSION](https://github.com/NVIDIA/nccl-tests)   | `2.13.9`    |
```

**Comment by sean-smith** [comment]
Looks almost ready. Just need the section on DL AMI NCCL test usage.

**Comment by mhuguesaws** [comment]
@sean-smith please review

**Comment by mhuguesaws** [comment]
@sean-smith anything else?

**Comment by mhuguesaws** [comment]
Yes.

**Review by sean-smith — APPROVED** [review]
LGTM

**Pull request merged**